File: src/qontinui/extraction/static/typescript/parser.js
#!/usr/bin/env node

import { readFileSync } from 'fs';
import { parse as babelParse } from '@babel/parser';
import traverse from '@babel/traverse';
import ts from 'typescript';

/**
 * QontinUI TypeScript/JavaScript Parser
 *
 * Parses TypeScript and JavaScript files to extract:
 * - Component definitions (function, class, arrow function)
 * - React hooks (useState, useReducer, useContext, etc.)
 * - Conditional rendering patterns
 * - Event handlers
 * - Import/export relationships
 * - JSX element hierarchy
 */

class TypeScriptParser {
  constructor() {
    this.results = {};
  }

  /**
   * Parse a single file and extract all relevant information
   */
  parseFile(filePath, extractTypes) {
    try {
      const sourceCode = readFileSync(filePath, 'utf-8');
      const fileExt = filePath.split('.').pop();

      // Determine parser options based on file extension
      const parserOptions = {
        sourceType: 'module',
        plugins: [
          'jsx',
          'typescript',
          'classProperties',
          'decorators-legacy',
          'dynamicImport',
          'exportDefaultFrom',
          'exportNamespaceFrom',
          'objectRestSpread',
          'optionalChaining',
          'nullishCoalescingOperator',
        ],
      };

      let ast;
      try {
        ast = babelParse(sourceCode, parserOptions);
      } catch (parseError) {
        // If Babel fails, try with more permissive settings
        console.error(`Babel parse error for ${filePath}: ${parseError.message}`, { file: 'stderr' });
        return {
          error: parseError.message,
          components: [],
          state_variables: [],
          conditional_renders: [],
          event_handlers: [],
          imports: [],
          exports: [],
        };
      }

      const fileResult = {
        components: [],
        state_variables: [],
        conditional_renders: [],
        event_handlers: [],
        imports: [],
        exports: [],
        jsx_elements: [],
      };

      // Track scope for better context
      const scopeStack = [];
      let currentComponent = null;

      traverse.default(ast, {
        // Extract imports
        ImportDeclaration(path) {
          if (!extractTypes.includes('imports')) return;

          const importInfo = {
            source: path.node.source.value,
            specifiers: path.node.specifiers.map(spec => {
              if (spec.type === 'ImportDefaultSpecifier') {
                return { type: 'default', name: spec.local.name };
              } else if (spec.type === 'ImportNamespaceSpecifier') {
                return { type: 'namespace', name: spec.local.name };
              } else {
                return {
                  type: 'named',
                  name: spec.local.name,
                  imported: spec.imported.name,
                };
              }
            }),
            line: path.node.loc?.start.line || 0,
          };
          fileResult.imports.push(importInfo);
        },

        // Extract exports
        ExportNamedDeclaration(path) {
          if (!extractTypes.includes('imports')) return;

          if (path.node.declaration) {
            const { declaration } = path.node;
            if (declaration.type === 'VariableDeclaration') {
              declaration.declarations.forEach(decl => {
                fileResult.exports.push({
                  type: 'named',
                  name: decl.id.name,
                  line: path.node.loc?.start.line || 0,
                });
              });
            } else if (declaration.id) {
              fileResult.exports.push({
                type: 'named',
                name: declaration.id.name,
                line: path.node.loc?.start.line || 0,
              });
            }
          }
        },

        ExportDefaultDeclaration(path) {
          if (!extractTypes.includes('imports')) return;

          const name = path.node.declaration.id?.name ||
                      path.node.declaration.name ||
                      'default';
          fileResult.exports.push({
            type: 'default',
            name,
            line: path.node.loc?.start.line || 0,
          });
        },

        // Extract function components
        FunctionDeclaration(path) {
          if (!extractTypes.includes('components')) return;

          const component = this.extractFunctionComponent(path);
          if (component) {
            fileResult.components.push(component);
            currentComponent = component;
          }
        },

        ArrowFunctionExpression(path) {
          if (!extractTypes.includes('components')) return;

          // Check if this is assigned to a variable (component declaration)
          if (path.parent.type === 'VariableDeclarator') {
            const component = this.extractArrowComponent(path);
            if (component) {
              fileResult.components.push(component);
              currentComponent = component;
            }
          }
        },

        // Extract class components
        ClassDeclaration(path) {
          if (!extractTypes.includes('components')) return;

          const component = this.extractClassComponent(path);
          if (component) {
            fileResult.components.push(component);
            currentComponent = component;
          }
        },

        // Extract React hooks (useState, useReducer, useContext, etc.)
        CallExpression(path) {
          const calleeName = this.getCalleeName(path.node.callee);

          // Extract state hooks
          if (extractTypes.includes('state')) {
            if (calleeName === 'useState' || calleeName === 'useReducer' ||
                calleeName === 'useContext' || calleeName === 'useRef' ||
                calleeName === 'useMemo' || calleeName === 'useCallback') {
              const stateVar = this.extractHook(path, calleeName);
              if (stateVar) {
                fileResult.state_variables.push(stateVar);
              }
            }
          }

          // Extract event handlers
          if (extractTypes.includes('handlers')) {
            // Look for common event handler patterns
            if (path.parent.type === 'JSXExpressionContainer' &&
                path.parentPath.parent.type === 'JSXAttribute') {
              const attrName = path.parentPath.parent.name.name;
              if (attrName && attrName.startsWith('on')) {
                const handler = this.extractEventHandler(path, attrName);
                if (handler) {
                  fileResult.event_handlers.push(handler);
                }
              }
            }
          }
        },

        // Extract conditional rendering patterns
        JSXElement(path) {
          if (!extractTypes.includes('conditionals') && !extractTypes.includes('components')) return;

          // Extract JSX hierarchy for component analysis
          if (extractTypes.includes('components')) {
            const element = this.extractJSXElement(path);
            if (element) {
              fileResult.jsx_elements.push(element);
            }
          }
        },

        LogicalExpression(path) {
          if (!extractTypes.includes('conditionals')) return;

          // Detect {condition && <Component />} pattern
          if (path.node.operator === '&&' && this.isJSXExpression(path.node.right)) {
            const conditional = {
              condition: this.getConditionString(path.node.left),
              line: path.node.loc?.start.line || 0,
              pattern: 'AND',
              renders: this.getRenderedComponents(path.node.right),
            };
            fileResult.conditional_renders.push(conditional);
          }
        },

        ConditionalExpression(path) {
          if (!extractTypes.includes('conditionals')) return;

          // Detect ternary operator: condition ? <A /> : <B />
          if (this.isJSXExpression(path.node.consequent) ||
              this.isJSXExpression(path.node.alternate)) {
            const conditional = {
              condition: this.getConditionString(path.node.test),
              line: path.node.loc?.start.line || 0,
              pattern: 'TERNARY',
              renders_true: this.getRenderedComponents(path.node.consequent),
              renders_false: this.getRenderedComponents(path.node.alternate),
            };
            fileResult.conditional_renders.push(conditional);
          }
        },

        IfStatement(path) {
          if (!extractTypes.includes('conditionals')) return;

          // Detect early returns: if (condition) return <Component />
          if (path.node.consequent.type === 'ReturnStatement' &&
              this.isJSXExpression(path.node.consequent.argument)) {
            const conditional = {
              condition: this.getConditionString(path.node.test),
              line: path.node.loc?.start.line || 0,
              pattern: 'EARLY_RETURN',
              renders: this.getRenderedComponents(path.node.consequent.argument),
            };
            fileResult.conditional_renders.push(conditional);
          }
        },
      });

      return fileResult;
    } catch (error) {
      console.error(`Error parsing ${filePath}: ${error.message}`, { file: 'stderr' });
      return {
        error: error.message,
        components: [],
        state_variables: [],
        conditional_renders: [],
        event_handlers: [],
        imports: [],
        exports: [],
      };
    }
  }

  /**
   * Extract function component information
   */
  extractFunctionComponent(path) {
    const name = path.node.id?.name;
    if (!name) return null;

    // Check if it returns JSX
    let returnsJSX = false;
    path.traverse({
      ReturnStatement(returnPath) {
        if (this.isJSXExpression(returnPath.node.argument)) {
          returnsJSX = true;
          returnPath.stop();
        }
      },
    });

    if (!returnsJSX && !this.isCapitalized(name)) return null;

    const props = this.extractProps(path.node.params);
    const children = this.extractChildComponents(path);

    return {
      name,
      type: 'function',
      line: path.node.loc?.start.line || 0,
      props,
      children,
      returns_jsx: returnsJSX,
    };
  }

  /**
   * Extract arrow function component information
   */
  extractArrowComponent(path) {
    const name = path.parent.id?.name;
    if (!name) return null;

    let returnsJSX = false;

    // Check direct JSX return
    if (this.isJSXExpression(path.node.body)) {
      returnsJSX = true;
    } else {
      // Check for JSX in return statements
      path.traverse({
        ReturnStatement(returnPath) {
          if (this.isJSXExpression(returnPath.node.argument)) {
            returnsJSX = true;
            returnPath.stop();
          }
        },
      });
    }

    if (!returnsJSX && !this.isCapitalized(name)) return null;

    const props = this.extractProps(path.node.params);
    const children = this.extractChildComponents(path);

    return {
      name,
      type: 'arrow_function',
      line: path.node.loc?.start.line || 0,
      props,
      children,
      returns_jsx: returnsJSX,
    };
  }

  /**
   * Extract class component information
   */
  extractClassComponent(path) {
    const name = path.node.id?.name;
    if (!name) return null;

    // Check if it extends React.Component or Component
    const extendsReact = path.node.superClass && (
      path.node.superClass.name === 'Component' ||
      path.node.superClass.name === 'PureComponent' ||
      (path.node.superClass.object?.name === 'React' &&
       (path.node.superClass.property?.name === 'Component' ||
        path.node.superClass.property?.name === 'PureComponent'))
    );

    if (!extendsReact && !this.isCapitalized(name)) return null;

    const children = this.extractChildComponents(path);

    return {
      name,
      type: 'class',
      line: path.node.loc?.start.line || 0,
      props: [],
      children,
      extends: extendsReact ? 'React.Component' : 'unknown',
    };
  }

  /**
   * Extract hook information
   */
  extractHook(path, hookName) {
    const parent = path.parent;

    let varName = null;
    let setterName = null;
    let initialValue = null;

    // Handle destructured assignment: const [state, setState] = useState(...)
    if (parent.type === 'VariableDeclarator' && parent.id.type === 'ArrayPattern') {
      const elements = parent.id.elements;
      varName = elements[0]?.name || null;
      setterName = elements[1]?.name || null;
    } else if (parent.type === 'VariableDeclarator' && parent.id.type === 'Identifier') {
      // Handle non-destructured: const ref = useRef(...)
      varName = parent.id.name;
    }

    // Extract initial value
    if (path.node.arguments.length > 0) {
      initialValue = this.getValueString(path.node.arguments[0]);
    }

    return {
      name: varName,
      setter: setterName,
      hook: hookName,
      line: path.node.loc?.start.line || 0,
      initial_value: initialValue,
      type: this.inferType(initialValue),
    };
  }

  /**
   * Extract event handler information
   */
  extractEventHandler(path, eventName) {
    const handler = {
      event: eventName.replace(/^on/, '').toLowerCase(),
      line: path.node.loc?.start.line || 0,
      state_changes: [],
    };

    // Try to get handler name
    if (path.node.type === 'Identifier') {
      handler.name = path.node.name;
    } else if (path.node.type === 'ArrowFunctionExpression' ||
               path.node.type === 'FunctionExpression') {
      handler.name = 'inline';

      // Look for state setter calls inside the handler
      path.traverse({
        CallExpression(callPath) {
          const calleeName = this.getCalleeName(callPath.node.callee);
          if (calleeName && calleeName.startsWith('set')) {
            handler.state_changes.push(calleeName);
          }
        },
      });
    } else if (path.node.type === 'CallExpression') {
      handler.name = this.getCalleeName(path.node.callee);
    }

    return handler;
  }

  /**
   * Extract JSX element information
   */
  extractJSXElement(path) {
    const openingElement = path.node.openingElement;
    const name = this.getJSXElementName(openingElement.name);

    return {
      name,
      line: path.node.loc?.start.line || 0,
      props: openingElement.attributes.map(attr => {
        if (attr.type === 'JSXAttribute') {
          return {
            name: attr.name.name,
            value: attr.value ? this.getValueString(attr.value) : true,
          };
        }
        return null;
      }).filter(Boolean),
      self_closing: path.node.openingElement.selfClosing,
    };
  }

  /**
   * Extract component props from function parameters
   */
  extractProps(params) {
    if (!params || params.length === 0) return [];

    const firstParam = params[0];
    const props = [];

    if (firstParam.type === 'ObjectPattern') {
      firstParam.properties.forEach(prop => {
        if (prop.type === 'ObjectProperty') {
          props.push({
            name: prop.key.name,
            default: prop.value.type === 'AssignmentPattern' ?
                     this.getValueString(prop.value.right) : undefined,
          });
        }
      });
    } else if (firstParam.type === 'Identifier') {
      props.push({ name: firstParam.name });
    }

    return props;
  }

  /**
   * Extract child component names from JSX
   */
  extractChildComponents(path) {
    const children = new Set();

    path.traverse({
      JSXElement(jsxPath) {
        const name = this.getJSXElementName(jsxPath.node.openingElement.name);
        if (name && this.isCapitalized(name)) {
          children.add(name);
        }
      },
    });

    return Array.from(children);
  }

  /**
   * Get JSX element name
   */
  getJSXElementName(node) {
    if (!node) return null;

    if (node.type === 'JSXIdentifier') {
      return node.name;
    } else if (node.type === 'JSXMemberExpression') {
      return `${this.getJSXElementName(node.object)}.${node.property.name}`;
    }
    return null;
  }

  /**
   * Get callee name from CallExpression
   */
  getCalleeName(node) {
    if (!node) return null;

    if (node.type === 'Identifier') {
      return node.name;
    } else if (node.type === 'MemberExpression') {
      return node.property.name;
    }
    return null;
  }

  /**
   * Check if expression is JSX
   */
  isJSXExpression(node) {
    if (!node) return false;
    return node.type === 'JSXElement' || node.type === 'JSXFragment';
  }

  /**
   * Get condition as string
   */
  getConditionString(node) {
    if (!node) return '';

    if (node.type === 'Identifier') {
      return node.name;
    } else if (node.type === 'MemberExpression') {
      return `${this.getConditionString(node.object)}.${node.property.name}`;
    } else if (node.type === 'UnaryExpression') {
      return `${node.operator}${this.getConditionString(node.argument)}`;
    } else if (node.type === 'BinaryExpression' || node.type === 'LogicalExpression') {
      return `${this.getConditionString(node.left)} ${node.operator} ${this.getConditionString(node.right)}`;
    }

    return 'complex_condition';
  }

  /**
   * Get rendered components from JSX expression
   */
  getRenderedComponents(node) {
    if (!node) return [];

    const components = [];

    if (node.type === 'JSXElement') {
      const name = this.getJSXElementName(node.openingElement.name);
      if (name) components.push(name);
    } else if (node.type === 'JSXFragment') {
      components.push('Fragment');
    }

    return components;
  }

  /**
   * Get value as string
   */
  getValueString(node) {
    if (!node) return null;

    if (node.type === 'StringLiteral') {
      return node.value;
    } else if (node.type === 'NumericLiteral') {
      return String(node.value);
    } else if (node.type === 'BooleanLiteral') {
      return String(node.value);
    } else if (node.type === 'NullLiteral') {
      return 'null';
    } else if (node.type === 'Identifier') {
      return node.name;
    } else if (node.type === 'ArrayExpression') {
      return '[]';
    } else if (node.type === 'ObjectExpression') {
      return '{}';
    } else if (node.type === 'JSXExpressionContainer') {
      return this.getValueString(node.expression);
    }

    return 'unknown';
  }

  /**
   * Infer type from initial value
   */
  inferType(value) {
    if (value === null) return 'unknown';
    if (value === 'true' || value === 'false') return 'boolean';
    if (value === '[]') return 'array';
    if (value === '{}') return 'object';
    if (value === 'null') return 'null';
    if (!isNaN(Number(value))) return 'number';
    return 'string';
  }

  /**
   * Check if name is capitalized (likely a component)
   */
  isCapitalized(name) {
    return name && name[0] === name[0].toUpperCase();
  }

  /**
   * Main parse function
   */
  parse(config) {
    const { files, extract } = config;
    const results = {};

    for (const file of files) {
      results[file] = this.parseFile(file, extract);
    }

    return { files: results };
  }
}

/**
 * Main execution
 */
async function main() {
  try {
    // Read configuration from stdin
    const chunks = [];
    for await (const chunk of process.stdin) {
      chunks.push(chunk);
    }
    const input = Buffer.concat(chunks).toString('utf-8');

    if (!input.trim()) {
      throw new Error('No input configuration provided');
    }

    const config = JSON.parse(input);

    if (!config.files || !Array.isArray(config.files)) {
      throw new Error('Invalid configuration: "files" array is required');
    }

    if (!config.extract || !Array.isArray(config.extract)) {
      config.extract = ['components', 'state', 'conditionals', 'handlers', 'imports'];
    }

    const parser = new TypeScriptParser();
    const results = parser.parse(config);

    // Output results to stdout
    console.log(JSON.stringify(results, null, 2));
    process.exit(0);
  } catch (error) {
    console.error(`Fatal error: ${error.message}`, { file: 'stderr' });
    process.exit(1);
  }
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export { TypeScriptParser };

File: src/qontinui/extraction/static/typescript/parser.py
"""
Python wrapper for the Node.js TypeScript/JavaScript parser.

This module provides a Python interface to the TypeScript parser that uses
Node.js and Babel to analyze TypeScript and JavaScript source code.
"""

import asyncio
import json
import subprocess
import tempfile
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


@dataclass
class PropInfo:
    """Information about a component prop."""

    name: str
    default: str | None = None


@dataclass
class ComponentInfo:
    """Information about a React component."""

    name: str
    type: str  # 'function', 'arrow_function', or 'class'
    line: int
    props: list[PropInfo]
    children: list[str]  # Names of child components
    returns_jsx: bool = True
    extends: str | None = None  # For class components


@dataclass
class StateVariableInfo:
    """Information about a React state variable."""

    name: str | None
    hook: str  # 'useState', 'useReducer', 'useContext', etc.
    line: int
    initial_value: str | None = None
    type: str = "unknown"
    setter: str | None = None


@dataclass
class ConditionalRenderInfo:
    """Information about conditional rendering."""

    condition: str
    line: int
    pattern: str  # 'AND', 'TERNARY', 'EARLY_RETURN'
    renders: list[str] = field(default_factory=list)
    renders_true: list[str] = field(default_factory=list)
    renders_false: list[str] = field(default_factory=list)


@dataclass
class EventHandlerInfo:
    """Information about an event handler."""

    event: str  # 'click', 'change', etc.
    line: int
    name: str | None = None
    state_changes: list[str] = field(default_factory=list)


@dataclass
class ImportInfo:
    """Information about an import statement."""

    source: str
    specifiers: list[dict[str, str]]
    line: int


@dataclass
class ExportInfo:
    """Information about an export statement."""

    type: str  # 'named' or 'default'
    name: str
    line: int


@dataclass
class JSXElementInfo:
    """Information about a JSX element."""

    name: str
    line: int
    props: list[dict[str, Any]]
    self_closing: bool


@dataclass
class NavigationLinkInfo:
    """Information about a navigation link."""

    type: str  # 'link' or 'anchor'
    target: str  # The href destination
    line: int
    component: str | None = None  # The component containing this link


@dataclass
class FileParseResult:
    """Result of parsing a single file."""

    components: list[ComponentInfo] = field(default_factory=list)
    state_variables: list[StateVariableInfo] = field(default_factory=list)
    conditional_renders: list[ConditionalRenderInfo] = field(default_factory=list)
    event_handlers: list[EventHandlerInfo] = field(default_factory=list)
    imports: list[ImportInfo] = field(default_factory=list)
    exports: list[ExportInfo] = field(default_factory=list)
    jsx_elements: list[JSXElementInfo] = field(default_factory=list)
    navigation_links: list[NavigationLinkInfo] = field(default_factory=list)
    error: str | None = None

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for downstream processing."""
        from dataclasses import asdict

        return {
            "components": [asdict(c) for c in self.components],
            "state_variables": [asdict(s) for s in self.state_variables],
            "conditional_renders": [asdict(c) for c in self.conditional_renders],
            "event_handlers": [asdict(h) for h in self.event_handlers],
            "imports": [asdict(i) for i in self.imports],
            "exports": [asdict(e) for e in self.exports],
            "jsx_elements": [asdict(j) for j in self.jsx_elements],
            "navigation_links": [asdict(n) for n in self.navigation_links],
            "function_declarations": [
                {"name": c.name, "line": c.line, "async": False, "parameters": []}
                for c in self.components
            ],
            "error": self.error,
        }


@dataclass
class ParseResult:
    """Result of parsing multiple files."""

    files: dict[str, FileParseResult]
    errors: list[str] = field(default_factory=list)


class TypeScriptParser:
    """
    Python wrapper for the Node.js TypeScript parser.

    This class provides a Python interface to analyze TypeScript and JavaScript
    files using the Node.js-based parser that leverages Babel and the TypeScript
    compiler API.
    """

    def __init__(self, node_path: str = "node"):
        """
        Initialize the TypeScript parser.

        Args:
            node_path: Path to the Node.js executable (default: "node")
        """
        self.node_path = node_path
        self.parser_script = Path(__file__).parent / "parser.js"

        if not self.parser_script.exists():
            raise FileNotFoundError(f"Parser script not found at {self.parser_script}")

    async def parse_files(
        self,
        files: list[Path],
        extract: list[str] | None = None,
    ) -> ParseResult:
        """
        Parse TypeScript/JavaScript files and extract information.

        Args:
            files: List of file paths to parse
            extract: List of what to extract. Options:
                - 'components': Component definitions
                - 'state': React hooks and state
                - 'conditionals': Conditional rendering patterns
                - 'handlers': Event handlers
                - 'imports': Import/export statements

        Returns:
            ParseResult containing extracted information from all files

        Raises:
            subprocess.CalledProcessError: If the parser fails
            json.JSONDecodeError: If the parser output is invalid
        """
        if extract is None:
            extract = ["components", "state", "conditionals", "handlers", "imports"]

        # Use a temporary file to avoid stdout buffer limits
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as output_file:
            output_path = output_file.name

        try:
            config = {
                "files": [str(f.resolve()) for f in files],
                "extract": extract,
                "outputFile": output_path,
            }

            # Run the Node.js parser
            process = await asyncio.create_subprocess_exec(
                self.node_path,
                str(self.parser_script),
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )

            # Send configuration to stdin
            config_json = json.dumps(config).encode("utf-8")
            stdout, stderr = await process.communicate(input=config_json)

            if process.returncode != 0:
                error_msg = stderr.decode("utf-8") if stderr else "Unknown error"
                raise RuntimeError(
                    f"Parser failed with exit code {process.returncode}: {error_msg}"
                )

            # Read results from the output file
            with open(output_path, encoding="utf-8") as f:
                result_data = json.load(f)

            return self._convert_to_dataclasses(result_data)

        except subprocess.CalledProcessError as e:
            raise RuntimeError(f"Failed to run Node.js parser: {e}") from e
        except json.JSONDecodeError as e:
            raise RuntimeError(f"Failed to parse parser output: {e}") from e
        finally:
            # Clean up the temporary file
            try:
                Path(output_path).unlink()
            except Exception:
                pass

    async def parse_directory(
        self,
        directory: Path,
        patterns: list[str] | None = None,
        exclude: list[str] | None = None,
        extract: list[str] | None = None,
    ) -> ParseResult:
        """
        Parse all matching files in a directory.

        Args:
            directory: Directory to search for files
            patterns: File patterns to include (e.g., ['*.ts', '*.tsx'])
            exclude: Patterns to exclude (e.g., ['node_modules/**', '*.test.ts'])
            extract: List of what to extract (see parse_files)

        Returns:
            ParseResult containing extracted information from all matching files
        """
        if patterns is None:
            patterns = ["*.ts", "*.tsx", "*.js", "*.jsx"]

        if exclude is None:
            exclude = [
                "node_modules/**",
                "dist/**",
                "build/**",
                "*.test.ts",
                "*.test.tsx",
                "*.test.js",
                "*.test.jsx",
                "*.spec.ts",
                "*.spec.tsx",
                "*.spec.js",
                "*.spec.jsx",
            ]

        # Find all matching files
        files = []
        for pattern in patterns:
            for file_path in directory.rglob(pattern):
                # Check if file should be excluded
                should_exclude = False
                for exclude_pattern in exclude:
                    if file_path.match(exclude_pattern):
                        should_exclude = True
                        break

                if not should_exclude and file_path.is_file():
                    files.append(file_path)

        if not files:
            return ParseResult(files={})

        return await self.parse_files(files, extract)

    def parse_files_sync(
        self,
        files: list[Path],
        extract: list[str] | None = None,
    ) -> ParseResult:
        """
        Synchronous version of parse_files.

        Args:
            files: List of file paths to parse
            extract: List of what to extract (see parse_files)

        Returns:
            ParseResult containing extracted information from all files
        """
        return asyncio.run(self.parse_files(files, extract))

    def parse_directory_sync(
        self,
        directory: Path,
        patterns: list[str] | None = None,
        exclude: list[str] | None = None,
        extract: list[str] | None = None,
    ) -> ParseResult:
        """
        Synchronous version of parse_directory.

        Args:
            directory: Directory to search for files
            patterns: File patterns to include
            exclude: Patterns to exclude
            extract: List of what to extract (see parse_files)

        Returns:
            ParseResult containing extracted information from all matching files
        """
        return asyncio.run(self.parse_directory(directory, patterns, exclude, extract))

    async def parse_file(
        self,
        file_path: str,
        extract: list[str] | None = None,
    ) -> FileParseResult:
        """
        Parse a single TypeScript/JavaScript file.

        Args:
            file_path: Path to the file to parse
            extract: List of what to extract (see parse_files)

        Returns:
            FileParseResult containing extracted information from the file
        """
        path = Path(file_path)
        result = await self.parse_files([path], extract)

        # Return the FileParseResult for this file, or an empty one if not found
        return result.files.get(str(path.resolve()), FileParseResult())

    def parse_file_sync(
        self,
        file_path: str,
        extract: list[str] | None = None,
    ) -> FileParseResult:
        """
        Synchronous version of parse_file.

        Args:
            file_path: Path to the file to parse
            extract: List of what to extract (see parse_files)

        Returns:
            FileParseResult containing extracted information from the file
        """
        return asyncio.run(self.parse_file(file_path, extract))

    def _convert_to_dataclasses(self, result_data: dict) -> ParseResult:
        """
        Convert raw JSON result to typed dataclasses.

        Args:
            result_data: Raw JSON data from the parser

        Returns:
            ParseResult with properly typed data
        """
        files = {}
        errors = []

        for file_path, file_data in result_data.get("files", {}).items():
            # Check for file-level errors
            if "error" in file_data and file_data["error"]:
                errors.append(f"{file_path}: {file_data['error']}")

            # Convert components
            components = [
                ComponentInfo(
                    name=c["name"],
                    type=c["type"],
                    line=c["line"],
                    props=[PropInfo(**p) for p in c.get("props", [])],
                    children=c.get("children", []),
                    returns_jsx=c.get("returns_jsx", True),
                    extends=c.get("extends"),
                )
                for c in file_data.get("components", [])
            ]

            # Convert state variables
            state_variables = [
                StateVariableInfo(
                    name=s.get("name"),
                    hook=s["hook"],
                    line=s["line"],
                    initial_value=s.get("initial_value"),
                    type=s.get("type", "unknown"),
                    setter=s.get("setter"),
                )
                for s in file_data.get("state_variables", [])
            ]

            # Convert conditional renders
            conditional_renders = [
                ConditionalRenderInfo(
                    condition=c["condition"],
                    line=c["line"],
                    pattern=c["pattern"],
                    renders=c.get("renders", []),
                    renders_true=c.get("renders_true", []),
                    renders_false=c.get("renders_false", []),
                )
                for c in file_data.get("conditional_renders", [])
            ]

            # Convert event handlers
            event_handlers = [
                EventHandlerInfo(
                    event=h["event"],
                    line=h["line"],
                    name=h.get("name"),
                    state_changes=h.get("state_changes", []),
                )
                for h in file_data.get("event_handlers", [])
            ]

            # Convert imports
            imports = [
                ImportInfo(
                    source=i["source"],
                    specifiers=i["specifiers"],
                    line=i["line"],
                )
                for i in file_data.get("imports", [])
            ]

            # Convert exports
            exports = [
                ExportInfo(
                    type=e["type"],
                    name=e["name"],
                    line=e["line"],
                )
                for e in file_data.get("exports", [])
            ]

            # Convert JSX elements
            jsx_elements = [
                JSXElementInfo(
                    name=j["name"],
                    line=j["line"],
                    props=j.get("props", []),
                    self_closing=j.get("self_closing", False),
                )
                for j in file_data.get("jsx_elements", [])
            ]

            # Convert navigation links
            navigation_links = [
                NavigationLinkInfo(
                    type=n["type"],
                    target=n["target"],
                    line=n["line"],
                    component=n.get("component"),
                )
                for n in file_data.get("navigation_links", [])
            ]

            files[file_path] = FileParseResult(
                components=components,
                state_variables=state_variables,
                conditional_renders=conditional_renders,
                event_handlers=event_handlers,
                imports=imports,
                exports=exports,
                jsx_elements=jsx_elements,
                navigation_links=navigation_links,
                error=file_data.get("error"),
            )

        return ParseResult(files=files, errors=errors)


def create_parser(node_path: str = "node") -> TypeScriptParser:
    """
    Create a new TypeScript parser instance.

    Args:
        node_path: Path to the Node.js executable

    Returns:
        TypeScriptParser instance
    """
    return TypeScriptParser(node_path)
